with_cfg/private/args.bzl

```diff
@@ -0,0 +1,166 @@
+
+1. The way locations expand in "args" depends on the configuration of the
+   underlying target, which can be affected by transitions both by with_cfg.bzl
+   and the rule itself. We thus need to obtain the expanded locations from the
+   rule context instead of e.g. rewriting labels to helper targets during the
+   loading phase, which would miss the effect of transitions applied by the rule
+   itself.
+   We use a non-recursing aspect on the original rule to collect the files
+   corresponding to labels mentioned in location expansion in "args" and provide
+   them to the frontend rule via output groups named after the user-provided
+   label strings. As a consequence, we also have to set the "args" attribute on
+   the underlying target so that the aspect can see it - aspect attributes are
+   limited to integers and string enums.
+2. "args" can only be set at load time via the magic attribute and the only way
+   to get analysis time information into it is via the hard-coded location
+   expansion applied to it in
+   https://github.com/bazelbuild/bazel/blob/9425e365ebf921d4286fcf159b429e38f6b0a48f/src/main/java/com/google/devtools/build/lib/analysis/RunfilesSupport.java#L525
+   We create a helper `filegroup` target for each unique label string mentioned
+   in a location expansion expression in "args", add it to the (otherwise
+   unused) `data` attribute of the frontend and rewrite the label strings to
+   instead point to the `filegroup` targets.
+3. `native.package_relative_label` is not available in the analysis phase and
+   rule or aspect implementation logic can't see the labels of alias targets,
+   so the only way to reliably match user-provided labels to the files they
+   represent is to query `ctx.expand_location("$(execpaths ...)")` and manually
+   collect the files with the given paths.
+"""
+
+load(":rewrite.bzl", "rewrite_locations_in_attr", "rewrite_locations_in_single_value")
+
+visibility("private")
+
+def rewrite_args(name, args, make_filegroup):
+    # type: (string, list[string], Any) -> tuple[list[string], list[Label]]
+    seen_labels = {}
+    filegroup_labels = []
+
+    def rewrite_label(label_string):
+        # type: (string) -> string
+        escaped_label = _escape_label(label_string)
+        filegroup_name = name + "__args__" + escaped_label
+        if label_string not in seen_labels:
+            seen_labels[label_string] = None
+            make_filegroup(
+                name = filegroup_name,
+                srcs = [":" + name],
+                output_group = _OUTPUT_GROUPS_PREFIX + label_string,
+            )
+            filegroup_labels.append(native.package_relative_label(filegroup_name))
+        return ":" + filegroup_name
+
+    return rewrite_locations_in_attr(args, rewrite_label), filegroup_labels
+
+def _args_aspect_impl(target, ctx):
+    # type: (Target, ctx) -> list[Provider]
+    if not hasattr(ctx.rule.attr, "args"):
+        return []
+
+    execpaths_expansion_to_files = {}
```

Comment:
@keith It's not pretty, but you may be interested in this trick to match user-provided location expansions to the files they refer to based on the Slack question you asked a few weeks ago.